Introdução para (hu)manos (#8)
nixosbrasil/onboard: src/0-introducao.md

# Introdução

> ## Nix o gerenciador de pacotes

Reviewer: ```suggestion
## O gerenciador de pacotes Nix
```

> **[Nix](https://nixos.org/explore.html)**<sup>🇬🇧</sup> é um gerenciador de pacotes de software como dpkg/apt do Debian ou rpm do RedHat. Entretanto, assim como AppImage, Flatpack e Snap, pode ser instalado em *qualquer distribuição Linux*, e também em *Android*, *MacOS* ou *WSL*.

Reviewer: Essas bandeirinhas são desnecessárias e deixam o texto um tanto repulsivo.

Author: Eu queria não ter bandeirinhas e apontar para links internos, mas ao mesmo tempo colocar um link para algo em ingles gratuitamente não gostei, e também não queria deixar sem link nenhum :'(

Da umas ideais aí ;-)

Reviewer: Já que é assim, abre uma seção no fim do texto, neste estilo:

```markdown
# Links / Referências
- [NixOS](https://nixos.org)
- [Homebrew](https://brew.sh)
```

> **[Nix](https://nixos.org/explore.html)**<sup>🇬🇧</sup> é um gerenciador de pacotes de software como dpkg/apt do Debian ou rpm do RedHat. Entretanto, assim como AppImage, Flatpack e Snap, pode ser instalado em *qualquer distribuição Linux*, e também em *Android*, *MacOS* ou *WSL*.

Reviewer: ```suggestion
**[Nix](https://nixos.org/explore.html)**<sup>🇬🇧</sup> é um gerenciador de pacotes, similar em conceito ao dpkg/apt (Debian), ao rpm (RedHat e ao Portage do Gentoo. Entretanto, de maneira similar ao [Homebrew](https://brew.sh/) ele pode ser instalado em qualquer distribuição *Linux*, e também em sistemas *Android*, *MacOS* e Windows (mediante *WSL*).
```

O Nix não é semelhante ao Flatpak, o conceito é bem diferente. Flatpak e amigos é algo mais próximo daqueles instaladores NSIS do mundo Windows.

Author: Flatpak e amigos não são pacotes que dá apra instalar em conjunto com os pacotes de sistema sem zoar coisa toda ao contrario de se eu juntasse dpkg, rpm,  portage e pacman?
O fato deles também terem uma interface melhor para usuários iniciantes não tira-lhes o merito.

Reviewer: Ser multiplataforma (ou, no caso, multidistro) não é o distintivo do flatpak.

E ninguém que vê o Nix olha e pensa "nossa, isso é igualzinho aos instaladores next-next-next-finish do Windows". Isso é o distintivo do flatpak.

E eu não estou falando de amigabilidade, mas do conceito mesmo. O Synaptic fornece uma interface bonitinha ao  APT mas isso não faz  o Synaptic "semelhante ao flatpak".

Author: Pois tenho mesma percepção sobre tua sugestão do homebrew, só consigo pensar em: 'olha um gerenciador de pacote para linha de comando do Mac', excluindo o isolamento do pacote, que no Mac era meio que padrão usando o 'gerenciador'' (arrastar o pacote para algum lugar), só não tinha repositório, mas repositório apt-get, yum, já tinham e sempre me fez pensar em homebrew como um apt-get/dpkg para que não tem um Unix decente.

Qualquer distro popular hoje já tem instalador next-next-finish, e já tinham muito antes do Flatpack e afins, se eles só existissem para isso, eles nem existiriam. O que eles realmente oferecem é:
- repositório único para _qualquer distro_ (nixpkgs)
- garantia de que o pacote vai funcionar em _qualquer distro_ (nix drv)

Mas popularmente um caso em que tua ideia se aplica a Flatpak e afins, é com Steamdeck, "de quero fazer algo amigável para o usuário, não tenho nenhum histórico/apego/bagagem de dpkg/rpm/pacman para carregar, e meu negócio é meu próprio ~gerenciador~ marketplace, para que deixar 3 interfaces?". Hipótese, pois não tenho para ter certeza, de que um usuário do steamdeck vai achar que Flatpak, ou seja lá o que usem, é o 'playstore' do linux.

E o no mundo dos servidores todo mundo perdeu para ~docker~ containers

> Seu principal foco é que os pacotes sejam 100% reproduzíveis, para isso isola a compilação do pacote, não permitindo acesso à internet, bibliotecas ou ferramentas externas que não possuam versão (hash) ou não estejam declaradas como dependência. 

Reviewer: ```suggestion
Nix tem um foco em _reproducibilidade_. Para isso, ele toma uma série de medidas, entre elas:

- separa a aquisição de fontes (código-fonte propriamente dito, _patches_ etc.) das demais fases de construção do pacote;
- exige verificação (mediante _hash sum_) das fontes adquiridas;
- impede o acesso à Internet durante a compilação do pacote;
- impede o acesso a arquivos e programas que não foram explicitamente requisitados como dependências;
```

Author: __Acho__ que _reproducibilidade_ devia ser _reprodutibilidade_

Reviewer: É vero.

> Também tenta isolar as dependências no momento da execução, por isso ao invés de usar as dependências de um diretório único compartilhado como */lib* usa vários diretórios com hash das dependencias */nix/store/HASH-DEPENDENCIA/lib*, permitindo diferentes versões da mesma dependência, trocando conflitos por duplicidade.

Reviewer: ```suggestion
Gerenciadores de pacotes tradicionais instalam seus pacotes diretamente na raiz do sistema, gerando arquivos espalhados em diversos diretórios.
Por outro lado, Nix instala cada pacote em um diretório dedicado, por padrão dentro de `/nix/store/`, em um formato similar a `/nix/store/<hash>-<nome do pacote>/{bin, lib, share, . . . }`
```

> A reproduzibilidade e o isolamento, permitem que qualquer máquina com Nix e SSH sejam utilizadas para gerar os pacotes, permitindo escalar a quantidade de empacotadores e deixar que outras maquinas mais potentes produzam o pacote caso seja um processo intensivo, como a compilação de código.
> 
> Pode have ainda um terceiro servidor especializado em hospedar os pacotes, o padrão é [cache.nixos.org](https://cache.nixos.org)<sup>🇬🇧</sup> que possui cache de boa parte pacotes de nixpkgs, porém, é possível configurar teu próprio utilizando outra máquina com SSH, HTTP ou API do S3.

Reviewer: ```suggestion

Este conjunto de medidas, bem como muitas outras, garante a reprodutibilidade da compilação. Em razão disso, é possível acelerar a compilação de pacotes ao mantê-los em um servidor de cache.
Por padrão, o projeto NixOS fornece o servidor [cache.nixos.org](https://cache.nixos.org),mas é perfeitamente possível configurar e utilizar outros.
```

Nix tem um foco em _reproducibilidade_. Para isso, ele toma uma série de medidas, entre elas:

- separa a aquisição de fontes (código-fonte propriamente dito, _patches_ etc.) das demais fases de construção do pacote;
- exige verificação (mediante _hash sum_) das fontes adquiridas;
- impede o acesso à Internet durante a compilação do pacote;
- impede o acesso a arquivos e programas que não foram explicitamente requisitados como dependências;

Gerenciadores de pacotes tradicionais instalam seus pacotes diretamente na raiz do sistema, gerando arquivos espalhados em diversos diretórios.
Por outro lado, Nix instala cada pacote em um diretório dedicado, por padrão dentro de `/nix/store/`, em um formato similar a `/nix/store/<hash>-<nome do pacote>/{bin, lib, share, . . . }`

Este conjunto de medidas, bem como muitas outras, garante a reprodutibilidade da compilação. Em razão disso, é possível acelerar a compilação de pacotes ao mantê-los em um servidor de cache.

> Por padrão, o projeto NixOS fornece o servidor [cache.nixos.org](https://cache.nixos.org),mas é perfeitamente possível configurar e utilizar outros.

Reviewer: ```suggestion
Por padrão, o projeto NixOS fornece o servidor [cache.nixos.org](https://cache.nixos.org), mas é perfeitamente possível configurar e utilizar outros.
```

> ## Nix a linguagem

Reviewer: ```suggestion
## A Linguagem Nix
```

> Para cada pacote são necessárias informações como nome, origem do código-fonte (caminho ou endereço online), versão (hash), dependências e instruções de compilação ou empacotamento. 
> 
> Essas informações devem ser descritas geralmente em um arquivo `default.nix` usando a linguagem programação de domínio especifico do Nix, também chamada **Nix** ou **Nix DSL**, é parecida com JSON, mas tem funções, variáveis pode importar outros arquivos Nix.

Reviewer: Acho que dá para deixar isso mais limpo:

```suggestion

Nix utiliza uma linguagem de programação própria para descrever o processo de implantação (_deploy_) de software.

Esta linguagem usa uma abordagem funcional semelhante a (Haskell)[https://www.haskell.org/], mas com uma sintaxe própria inspirada em [JSON](https://www.json.org/json-pt.html).
```

Author: "Eu até estava pensando em instalar, mas você falou funcional e haskell, deve ser chato difícil" - Pensa Joaozinho de 19 anos em seu Ubuntu.

Author: Se a gente falar derivation nunca mais vai ver o cara na vida :bomb: :rofl:

Reviewer: Desculpa, mas desde que Java começou a incluir elementos funcionais na linguagem, isso deixou de ser coisa de nerd leitor de paper do arxiv.

Ubuntu? Por favor, a distro do momento para novatos é o Pop!OS e o Linux Mint, se atualiza.

Enfim, isso é só uma curiosidade sobre as origens da linguagem. Ou você acha mermu que esse hipotético Joãozinho Ubunteiro de 19 anos sabe o que é JSON?
E se souber, é bem capaz que tenha visto programação funcional em Javascript, haha! 

Ninguém está sugerindo ser necessário estudar Haskell para usar Nix, não mais que estudar Modula para entender Python.

Reviewer: \> deriveishoun assusta as pessoa

\> DSL
 
🤦

Author: Só se for 39% (Pop + Mint) :rofl: 
![Image](https://truelist.co/wp-content/uploads/2022/01/Top-Linux-Subcategories-by-Market_Share.jpg)

Author: > Isso deixou de ser coisa de nerd leitor de paper do arxiv.

[ Citation needed ]

> Joãozinho Ubunteiro de 19 anos sabe o que é JSON?

Preferia que ele não precisasse saber o que é JSON para usar Nix,
Mas como é muito orientado a configuração, nosso mínimo é:
Abrir um arquivo num editor de texto, ter noção de que aquilo respeita uma estrutura.
Em 24/01/2022 fiz a seguinte estatística nada ciêntifica (busquei no github por quantidade de arquivos)

![image](https://github.com/nixosbrasil/onboard/assets/863299/f9f8c657-2b35-4d7e-a88b-c92e91135f3c)
![image](https://github.com/nixosbrasil/onboard/assets/863299/47bd5e76-3957-423c-ba6a-1b8054dbf084)

> Ninguém está sugerindo ser necessário estudar Haskell para usar Nix, não mais que estudar Modula para entender Python.

![image](https://github.com/nixosbrasil/onboard/assets/863299/3cea9bdd-5a7d-42b5-a9be-0cbdcc7403ec)

Se colocar Haskel nessa imagem aí que o cara desiste mesmo

Reviewer: > [citation needed]

https://opensource.com/article/17/6/functional-javascript

https://www.amazon.com.br/Functional-JavaScript-Steve-Vinoski/dp/1449360726

https://medium.com/@mvalho/java-8-functional-interfaces-tornando-o-java-mais-legal-72401462d0e2

2018. Tipo, século passado em tempo de programador.

E Java 8 era novidade na época. Hoje em dia estamos no Java 20.

> Só se for 39% (Pop + Mint)

O [fonte](https://www.enterpriseappstoday.com/stats/linux-statistics.html) fala de websites. Mas que seja.

E, das langs aí em cima, acho que só C e HTML não têm suporte a paradigmas funcionais.

Author: > 2018. Tipo, século passado em tempo de programador.

Nix tem 20 anos, dá alguns séculos nessa linha de pensamento, e não é popular

Quando chega gente nova e pergunta alguma coisa a gente fala:
Nóis: - Sabe haskell...?
Newbie: - não
Nóis: - Ok, mas sabe OCaml?
Newbie: - não
Nóis: - Lisp, Erlang, F#, Clojure, Scala...
Newbie: - não
Nóis: -
![image](https://github.com/nixosbrasil/onboard/assets/863299/214b7569-0c9e-4b02-bb30-89e1c592827f)

Reviewer: > Nix tem 20 anos, dá alguns séculos nessa linha de pensamento, e não é popular

Eu apontei que Java - uma linguagem com [programadores já velhotes](https://www.youtube.com/watch?v=fSX86nxuEBs) implementou paradigmas funcionais há um bom tempo.

> Quando chega gente nova e pergunta alguma coisa a gente fala:

\- Já viu Python?
\- Já
\- Manja o lambda do Python? (ou o list comprehension)
\- Um pouco
\- Pois bem, isso é programação funcional

ACCESS GRANTED
![image](https://github.com/nixosbrasil/onboard/assets/5954806/6a620926-6f2d-4ac2-9b8b-978ffdab95ed)

Mas não vou perseguir isso.

Author: Lembrei disso

@lucasew, toda vez que a gente fala derivation:
https://www.instagram.com/reel/CwoLlzYpuyt/?igshid=MzRlODBiNWFlZA==

> ## Nixpkgs a coleção de pacotes

Reviewer: ```suggestion
## Nixpkgs
```

Author: Veja bem a pessoa não sabe o que nixpkgs é, queria deixar claro nos subtitulos que tem
Nix, o gerenciador de pacote (standalone)
Nix, a linguagem (em que se define os pacotes)
Nixpkgs, o repositorio padrão
NixOS, o sistema operacional que usa tudo a cima

Reviewer: Headers não são feitos para isso. Para isso se usa o primeiro parágrafo depois do header.
Headers servem para estruturação lógica.

> Existe uma coleção comunitária com mais de 80 mil pacotes Nix, que está configurada por padrão, chamada **[nixpkgs](https://search.nixos.org)**, às vezes referenciada apenas como **pkgs**. 

Reviewer: ```suggestion
[Nixpkgs](https://github.com/NixOS/nixpkgs) é uma coleção de mais de 80.000 pacotes que podem ser instalados pelo gerenciador Nix.
```

Author: Gostei do texto, mas é para iniciantes, apontar para o codigo é tipo chegar e perguntar se querem ver o [relógio preto](https://www.youtube.com/watch?v=cPO9wa5C2f4)

Reviewer: Justo. Mas uma search engine não é lá a melhor apresentação.

Que seja então o manual: https://nixos.org/manual/nixpkgs/stable/

Author: > Thus it’s mainly for packagers and developers who want to add packages to Nixpkgs.

:disappointed: 
A pessoa está na introdução, ainda acho o search menos obtuso.

Na verdade eu queria um nixpkgs guide for nix users e isso aí é mais para um API Guide, Deployment Guide, se fosse seguindo o exemplo abaixo

![image](https://github.com/nixosbrasil/onboard/assets/863299/1958862a-4f45-40d5-95ed-1f51edb6cc0b)
[source](https://www.cisco.com/c/en/us/td/docs/video/virtualized_video_processing/v2p_controller/3-3-3/release-notes/release-notes-for-v2pc-3-3-3_v2.html#pgfId-1722745)

Reviewer: >  Na verdade eu queria um nixpkgs guide for nix users 

Não temos nada oficial neste sentido. 

Ademais, o link precisa ser oficial, não _for humans_. Até porque este documento que você está editando é precisamente o link para humanos.

Author: E qual link oficial condiz mais com isso `é uma coleção de mais de 80.000 pacotes que podem ser instalados pelo gerenciador Nix.`?

Reviewer: Por mim, poderia simplesmente apontar a homepage nixos mesmo.
Afinal, tem uma issue exatamente sobre a falta de páginas dedicadas a explanar o Nixpkgs.

https://github.com/NixOS/nixos-homepage/issues/734

> Apesar de todos pacotes estarem configurados, devido à linguagem Nix só ler que for de fato utilizado, nenhum destes pacotes será instalado em sua máquina a menos que usados como dependência de algum outro *pacote* ou um *pacote de ambiente* que esteja instalado.

Reviewer: ```suggestion
```

Isso aqui mais atrapalha que ajuda. Afinal, qualquer gerenciador de pacotes decente funciona dessa forma, instalando somente o que foi requisitado.
Exceto talvez pelo Slackware, o gerenciamento automático de dependências também é parte do trabalho do gerenciador de pacotes.

Author: Concordo contigo,
Eu só queria deixar claro a parte lazy da linguagem nix, mas talvez seja muito avançado para deixar aqui

Reviewer: Você até poucos parágrafos atrás estava reclamando de citar - apenas _citar_ - Haskell.
Mas acha de boas descompactar todo o conceito de _linguagem de evaluação não-estrita preguiçosa_?

Vamos alinhar as expectativas aê...

Author: > It is said that there's a curse with Monads.. When you finally understand it, you'll lose the ability to explain it to others.

Não sou imune a maldição. ;-)

Por isso fico muito feliz e grato pelos teus reviews.

> ### NixOS sistema operacional como *pacote de ambiente*

Reviewer: ```suggestion
### NixOS
```

> **[NixOS](https://nixos.org/download.html#nixos-iso)** é um pacote Nix contendo todas as dependências para se iniciar o *sistema operacional linux*, tal como kernel, grub ou efi e outras como systemd, xorg ou wayland, e ferramentas de reinstalação.
> 
> As configurações (ie. quase tudo que costuma ir no /etc), são pacotes e para facilitar a criação dos pacotes de configuração, foi criado um framework também usando a linguagem Nix, chamado **[Nix Modules](https://search.nixos.org/options)**, que permite usuário validar e reutilizar configurações para, por exemplo, criação de usuários, instalação pacotes e serviços.
> 
> A reconfiguração de NixOS na realidade é a geração de um novo pacote, referenciando todos pacotes e configurações (inalteradas e alteradas) e que é reinstalado. Como o pacote anterior não é automaticamente descartado, lhe será permitido reinstalar a versão anterior para casos de falha com a versão nova.

Reviewer: ```suggestion
**[NixOS](https://nixos.org/)** é uma distribuição Linux de referência, construída com base no gerenciador Nix e nos pacotes do Nixpkgs.

NixOS é projetada como um conjunto de módulos que podem ser combinados pelo usuário, permitindo uma vasta gama de customizações, indo desde o gerenciador de boot e o kernel até desktops inteiros como o KDE Plasma e o Gnome.

Tal configuração não está limitada à simples instalação dos softwares, mas também se estende às suas configurações. Enquanto em distros tradicionais o usuário necessita editar arquivos espalhados pelo `/etc`, o sistema de módulos do NixOS também realiza esta configuração.

Grosso modo, o próprio NixOS é tratado como um pacote de software do ponto de vista do gerenciador Nix.

Uma propriedade interessante do sistema NixOS é o suporte à reversão (_rollback_), onde iterações anteriores (conhecidas como _gerações_) não são automaticamente apagadas, permanecendo disponíveis para serem utilizadas em caso de alguma falha na geração atual.

```

> ### HomeMananger configurações do usuário como *pacote de ambiente*

Reviewer: ```suggestion
### Home Mananger
```

> Utilizando o mesmo conceito do NixOS, **[HomeManager](https://github.com/nix-community/home-manager)** é um *pacote das configurações do usuário*, permitindo instalar pacotes ou configurações especificas para seu usuário, apesar de NixOS também permitir configurações por usuário, HomeManager pode ser instalado em outras distribuições, MacOS e WSL, além disto, possui algumas configurações únicas, sendo usado também em conjunto com NixOS.

Reviewer: ```suggestion
O **[HomeManager](https://nix-community.github.io/home-manager)** é outro sistema modular escrito em Nix, com a diferença que este é voltado para o usuário, em contraste com o NixOS, que é voltado para o sistema como um todo.

Enquanto o NixOS alavanca o Nixpkgs para o administrador do sistema operacional, o Home Manager alavanca o mesmo Nixpkgs para o usuário comum. 

A exemplo do seu irmão maior, o Home Manager também instala e configura os pacotes instalados, auxiliando na geração automatizada dos famigerados _dotfiles_.

Atualmente, o Home Manager conta com suporte ao NixOS, ao MacOS X e a virtualmente todas as distros Linux, o que o torna uma opção interessante para quem deseja utilizar softwares ainda não presentes na sua distro local.
```

Author: Ficou melhor, só para parte do 'usuário local' que ficou confusa :thinking: 
talvez se deixar só 'usuário'

Reviewer: A ideia seria separar o usuário do admin.
Mas realmente isso deveria ser óbvio.

Dá para reescrever ressaltando essa diferença, em especial o fato de ser possível instalar coisas sem ser `root`.

Reviewer: Atualizei esse diff

> ### Projetos como *pacote de ambiente*

Reviewer: ```suggestion
### Projetos que usam Nix
```

Author: Não sei se o título bate,
a ideia inicial era:
nix, onde tudo é um pacote
nixpkgs, coleção de pacotes
nixos, um pacote que é uma distro e configs são pacotes
nix-modules, uma mão na roda para gerar pacotes de config para NixOS
home-manager, um pacote que são as configs do usuário
mkShell, um pacote para projetos de dev, ops ou data science

'Projetos que usam' esta mais para: olha tem quem use.
Talvez, 'Gestão de projetos com Nix' :thinking:

Reviewer: Essa parte poderia ser então chamada de funcionalidades extras, e poderia ser bem mais enxuta.

Tal configuração não está limitada à simples instalação dos softwares, mas também se estende às suas configurações. Enquanto em distros tradicionais o usuário necessita editar arquivos espalhados pelo `/etc`, o sistema de módulos do NixOS também realiza esta configuração.

Grosso modo, o próprio NixOS é tratado como um pacote de software do ponto de vista do gerenciador Nix.

Uma propriedade interessante do sistema NixOS é o suporte à reversão (_rollback_), onde iterações anteriores (conhecidas como _gerações_) não são automaticamente apagadas, permanecendo disponíveis para serem utilizadas em caso de alguma falha na geração atual.

### HomeMananger configurações do usuário como *pacote de ambiente*

Utilizando o mesmo conceito do NixOS, **[HomeManager](https://github.com/nix-community/home-manager)** é um *pacote das configurações do usuário*, permitindo instalar pacotes ou configurações especificas para seu usuário, apesar de NixOS também permitir configurações por usuário, HomeManager pode ser instalado em outras distribuições, MacOS e WSL, além disto, possui algumas configurações únicas, sendo usado também em conjunto com NixOS.

### Projetos como *pacote de ambiente*

> Caso um desenvolvedor de software, cientista de dados, ou qualquer profissional que precise *ferramentas diferentes para diferentes projetos* do dia a dia, é possível criar um *ambiente para cada projeto*, atualmente exitem diversas ferramentas, algumas usando Nix Modules como [DevEnv](https://github.com/cachix/devenv) e [DevShell](https://github.com/numtide/devshell) outras usando apenas pacotes como [DevBox](https://github.com/jetpack-io/devbox) e [Flox](https://github.com/flox/flox). Também é possível criar teu próprio ambiente apenas com Nix e Nixpkgs criando um arquivo `shell.nix` contendo as definições do teu pacote de ambiente.

Reviewer: Note que boa parte desses projetos usa o Nix/Nixpkgs como base, mas não são oficiais. Uma rápida menção na seção de outros já basta.

> Existe ainda a possibilidade de iniciar ambiente diretamente usando o comando `nix-shell -p DEPENDENCIA`.

Reviewer: Esse aqui merecia um pouco mais de molho. Por exemplo, uma nix-shell rodando um programinha simples como  o cowsay...
